File: backend/src/main/kotlin/com/example/backend/controllers/SortieController.kt
package com.example.backend.controllers

import com.example.backend.dtos.CommandePageableCustomlDto
import com.example.backend.dtos.SortieDetailDto
import com.example.backend.models.Forme
import com.example.backend.models.ProduitDetail
import com.example.backend.services.FormeService
import com.example.backend.services.SortieStockService
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/sortie-stock")
class SortieController(private val sortieStockService: SortieStockService) {

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping
  fun getSortieStockPageable(
    @RequestParam(required = false) nomProduit: String?,
    @RequestParam(required = false) typeSortie: String?,
    @RequestParam(required = false) enRayonId: String?,
    @RequestParam(required = false) produitDetailId: String?,
    @RequestParam(required = false) search: String?,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "desc") direction: String
  ): ResponseEntity<Page<Map<String, Any?>>> {
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "dateSortie"))
    val enRayonIdLong = enRayonId?.toLongOrNull() ?: 0
    val produitDetailIdLong = produitDetailId?.toLongOrNull() ?: 0
    val result =
      sortieStockService.getSortieStockPageable(nomProduit, typeSortie, enRayonIdLong, produitDetailIdLong, pageable)
    return ResponseEntity.ok(result)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/product")
  fun getSortieStockPageableProductRange(
    @RequestParam(required = false) nomProduit: String?,
    @RequestParam(required = false) produitId: String?,
    @RequestParam(required = false) startDate: String?,
    @RequestParam(required = false) endDate: String?,
    @RequestParam(required = false) typeSortie: String?,
    @RequestParam(required = false) enRayonId: String?,
    @RequestParam(required = false) produitDetailId: String?,
    @RequestParam(required = false) search: String?,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "desc") direction: String
  ): ResponseEntity<CommandePageableCustomlDto> {
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "dateSortie"))
    val enRayonIdLong = enRayonId?.toLongOrNull() ?: 0
    val produitDetailIdLong = produitDetailId?.toLongOrNull() ?: 0
    val result = sortieStockService.getSortieStockPageableProductRange(
      nomProduit,
      produitId,
      startDate,
      endDate,
      typeSortie,
      enRayonIdLong,
      produitDetailIdLong,
      pageable
    )
    return ResponseEntity.ok(result)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/save")
  fun getSortieStockPageable(
    @RequestBody sortie: SortieDetailDto
  ): ResponseEntity<ProduitDetail> {
    val result = sortieStockService.addProduitDetail(sortie)
    return ResponseEntity.ok(result)
  }

}

File: backend_py/pharmaPython/app/api/v1/sortie_controller.py
from fastapi import APIRouter, Depends, Query, Body
from sqlalchemy.orm import Session
from typing import Any, Dict, Optional

from app.api.deps import get_db
from app.schemas.sortie_dto import SortieDetailDto
from app.services.sortie_stock_service import SortieStockService
from app.utility.jwt_authentication import jwt_authentication

router = APIRouter(prefix="/sortie-stock", tags=["Sorties de stock"],  dependencies=[Depends(jwt_authentication)],)

# GET /
@router.get("")
def get_sortie_stock_pageable(
  nomProduit: Optional[str] = Query(None),
  typeSortie: Optional[str] = Query(None),
  enRayonId: Optional[str] = Query(None),
  startDate: Optional[str] = Query(None),
  endDate: Optional[str] = Query(None),
  produitDetailId: Optional[str] = Query(None),
  search: Optional[str] = Query(None),
  page: int = Query(0, ge=0),
  size: int = Query(10, ge=1),
  sort: str = Query("id"),
  direction: str = Query("desc"),
  db: Session = Depends(get_db),
):
  """
  Émule Page<Map<...>> Spring :
  renvoie {content, totalElements, totalPages, pageSize, pageNumber}
  Tri effectif par 'dateSortie' DESC comme en Kotlin.
  """
  svc = SortieStockService(db)
  en_rayon_id = int(enRayonId) if (enRayonId and enRayonId.isdigit()) else 0
  produit_detail_id = int(produitDetailId) if (produitDetailId and produitDetailId.isdigit()) else 0

  result = svc.get_sortie_stock_pageable(
    nom_produit=nomProduit,
    type_sortie=typeSortie,
    en_rayon_id=en_rayon_id,
    produit_detail_id=produit_detail_id,
    page=page,
    size=size,
  )
  total = result.get("totalElements", 0)
  result.update({
    "totalPages": (total + size - 1) // size if size else 1,
    "pageSize": size,
    "pageNumber": page,
    "sort": "dateSortie",
    "direction": "DESC",
  })
  return result

# GET /product
@router.get("/product")
def get_sortie_stock_pageable_product_range(
  nomProduit: Optional[str] = Query(None),
  produitId: Optional[str] = Query(None),
  supprimer: Optional[str] = Query(None),
  startDate: Optional[str] = Query(None),
  endDate: Optional[str] = Query(None),
  typeSortie: Optional[str] = Query(None),
  enRayonId: Optional[str] = Query(None),
  produitDetailId: Optional[str] = Query(None),
  search: Optional[str] = Query(None),
  page: int = Query(0, ge=0),
  size: int = Query(10, ge=1),
  sort: str = Query("id"),
  direction: str = Query("desc"),
  db: Session = Depends(get_db),
):
  svc = SortieStockService(
    db)
  en_rayon_id = int(enRayonId) if (enRayonId and enRayonId.isdigit()) else 0
  produit_detail_id = int(produitDetailId) if (produitDetailId and produitDetailId.isdigit()) else 0

  dto = svc.get_sortie_stock_pageable_product_range(
    nom_produit=nomProduit,
    produit_id=produitId,
    supprimer=supprimer,
    start_date=startDate,
    end_date=endDate,
    type_sortie=typeSortie,
    en_rayon_id=en_rayon_id,
    produit_detail_id=produit_detail_id,
    page=page,
    size=size,
  )
  # dto de type CommandePageableCustomlDto côté Kotlin → dict Python
  dto["pageNumber"] = page
  dto["pageSize"] = size
  dto["sort"] = "date_sortie"
  dto["direction"] = "DESC"
  return dto

# POST /save
@router.post("/save")
def add_sortie_stock(sortie: SortieDetailDto = Body(...), db: Session = Depends(get_db)):
  """
  Équivalent du POST /save Kotlin : ajoute une sortie sur un ProduitDetail.
  """
  svc = SortieStockService(db)
  return svc.add_produit_detail(sortie)
